compiler/rustc_codegen_ssa/src/back/metadata.rs

```diff
@@ -172,6 +172,330 @@ pub(crate) fn create_object_file(sess: &Session) -> Option<write::Object<'static
             let e_flags = elf::EF_RISCV_RVC | elf::EF_RISCV_FLOAT_ABI_DOUBLE;
             e_flags
         }
+        Architecture::Avr => {
+            // Adapted from llvm-project/llvm/lib/target/AVR/AVRDevices.td
+            match sess.target.options.cpu.as_ref() {
```

Comment:
It looks like we only support atmega328 at the moment. For AVR we pass a hard-coded linker flag (`-mmcu=atmega328` for avr-unknown-gnu-atmega328) containing which AVR variant is used to the linker. I can imagine that it going out of sync with `-Ctarget-cpu` is a really bad idea. Instead it did probably need a separate target spec for each cpu. Especially if you can't mix a standard library compiled for one variant with one compiled for another variant. Maybe this `EF_AVR_MACH_*` constant could be specified to the target spec instead of using a giant match? Or maybe not? By the way, should we forbid `-Ctarget-cpu` for AVR?

Reply:
> It looks like we only support atmega328 at the moment

> Instead it did probably need a separate target spec for each cpu

Only atmega328 is in mainline rustc, but the Rust-AVR community supports many more using [custom target specs](https://github.com/Rahix/avr-hal/tree/main/avr-specs).

> For AVR we pass a hard-coded linker flag 

I am pretty sure that the `-mmcu` flag is not "hardcoded", it should be set based on the `cpu` target option, which is set in our target specs. In my experience, `-mmcu` does get set correctly for other CPUs.

> Maybe this `EF_AVR_MACH_*` constant could be specified to the target spec instead of using a giant match? 

Both AVR-GCC and LLVM accept the CPU name as a single flag (e.g. `-mmcu`) and use that to look up values like  `EF_AVR_MACH_*` internally (e.g. with the `AVRDevices.td` table that I mentioned in the comment).

That being said, being able to specify the ISA revision in the target spec is also a good idea. It would make maintenance in rustc a little easier, as rustc would only have to maintain the list of ISAs and not every CPU ever released.

(CC @Rahix - would it be acceptable to specify the ISA alongside the CPU in the target specs?)

>  should we forbid `-Ctarget-cpu` for AVR?

I do not see a reason for doing that.

Reply:
> I do not see a reason for doing that.

Does using say `-Ctarget-cpu=avr1` for one crate and `-Ctarget-cpu=avrtiny` for another crate work? If not, they have to be treated as separate targets without an overlapping set of allowed values for `-Ctarget-cpu`. The way that rustc checks for compatibility is by checking that the target name is identical. It assumes that linking crates with different `-Ctarget-cpu` values works fine and then runs on any cpu implementing the union of whatever features these `-Ctarget-cpu` values imply.

Reply:
For AVR specifically, the main usecase I can think of is picking a common-denominator ISA for building a "portable" library. I don't know whether that's actually useful or not. Probably not, since we currently specify each AVR CPU as its own target anyway.

And sure, if you misuse that, you may end up with a broken executable that doesn't work on your platform. That's true for any target. 

However, I don't think it would be productive to add more code to forbid it. It's not like it's a "footgun" that a developer would regularly encounter and accidentally set incorrectly without realizing.

Reply:
I was worried about say AVR2 cpu's not being able to run AVR1 code. That AVR1 cpu's can't run AVR2 code is obvious.

> It's not like it's a "footgun" that a developer would regularly encounter and accidentally set incorrectly without realizing.

You might not realize that the standard library is compiled for an incompatible cpu even if you set `-Ctarget-cpu` for your own crates.

In any case I guess it is fine to leave it as is.

Reply:
> (CC @Rahix - would it be acceptable to specify the ISA alongside the CPU in the target specs?)

I certainly wouldn't mind at all.  However both gcc and llvm also seem to maintain the list of MCUs so I think it is just as reasonable to do the same in rustc.